### PR TITLE
Add zed

### PR DESCRIPTION
**Pull request by charmoniumQ**

### CODE REVIEW

File: zed/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+SET(EXECUTABLE_OUTPUT_PATH ".")
+
+find_package(ZED 3 REQUIRED)
+find_package(OpenCV REQUIRED)
+find_package(CUDA ${ZED_CUDA_VERSION} EXACT REQUIRED)
+
+IF(NOT WIN32)
+    add_definitions(-Wno-format-extra-args)
+    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
+    SET(SPECIAL_OS_LIBS "pthread" "X11")
```
Comment (mhuzai) [inline zed/CMakeLists.txt:25]: Why does ZED need X11?

File: zed/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+link_directories(${ZED_LIBRARY_DIR})
+link_directories(${OpenCV_LIBRARY_DIRS})
+link_directories(${CUDA_LIBRARY_DIRS})
+
+FILE(GLOB_RECURSE SRC_FILES src/main.cpp)
+FILE(GLOB_RECURSE HDR_FILES include/*.hpp)
+
+ADD_EXECUTABLE(${execName} ${HDR_FILES} ${SRC_FILES})
+add_definitions(-std=c++17 -g -O3)
```
Comment (mhuzai) [inline zed/CMakeLists.txt:41]: Shouldn't `DCMAKE_BUILD_TYPE` take care of setting -g and -O3?

File: zed/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+if(COMMAND cmake_policy)
+	cmake_policy(SET CMP0003 OLD)
+	cmake_policy(SET CMP0015 OLD)
+endif(COMMAND cmake_policy)
+
+SET(EXECUTABLE_OUTPUT_PATH ".")
+
+find_package(ZED 3 REQUIRED)
+find_package(OpenCV REQUIRED)
+find_package(CUDA ${ZED_CUDA_VERSION} EXACT REQUIRED)
```
Comment (mhuzai) [inline zed/CMakeLists.txt:20]: This reminds me: ZED SDK needs to be part of `install_deps.sh` now. But this would be tricky because ZED SDK has very specific requirements for OS and CUDA version. So we'll have to detect which OS and CUDA the user already has. And speaking of CUDA, should we add that as a dependency as well or leave that up to the user? We have so far been assuming that the user already has it.

File: zed/src/main.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+  {
+    // Image setup
+    imageL_zed.alloc(image_size.width, image_size.height, MAT_TYPE::U8_C4, MEM::CPU);
+    imageR_zed.alloc(image_size.width, image_size.height, MAT_TYPE::U8_C4, MEM::CPU);
+
+    imageL_ocv = slMat2cvMat(imageL_zed);
+    imageR_ocv = slMat2cvMat(imageR_zed);
+
+		// experimental
+		runtime_parameters.enable_depth = false;
```
Comment (mhuzai) [inline zed/src/main.cpp:81]: @Hyjale , is this still required?

File: zed/src/main.cpp [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+      // Retrieve images
+      zedm->retrieveImage(imageL_zed, VIEW::LEFT, MEM::CPU, image_size);
+      zedm->retrieveImage(imageR_zed, VIEW::RIGHT, MEM::CPU, image_size);
+
+	  auto start_cpu_time  = thread_cpu_time();
+	  auto start_wall_time = std::chrono::high_resolution_clock::now();
+
+      // Conversion
+      // cv::cvtColor(imageL_ocv, grayL_ocv, CV_BGR2GRAY);
+      // cv::cvtColor(imageR_ocv, grayR_ocv, CV_BGR2GRAY);
```
Comment (mhuzai) [inline zed/src/main.cpp:118]: Shouldn't there be a corresponding change in OV where it does this conversion itself?

**Review by mhuzai — COMMENTED** [review]
Left a few comments but didn't actually test the code yet.

File: zed/src/main.cpp [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    time_type t;
+    ullong imu_time;
+
+	  std::size_t last_serial_no {0};
+};
+
+// This line makes the plugin importable by Spindle
+PLUGIN_MAIN(zed_imu_thread);
+
+int main(int argc, char **argv) { return 0; }
```
Comment (mhuzai) [inline zed/src/main.cpp:238]: Why is this required?

File: zed/src/main.cpp (hunk 4 above)
Comment (Hyjale) [reply zed/src/main.cpp:81]: Not required.

File: zed/CMakeLists.txt (hunk 3 above)
Comment (Hyjale) [reply zed/CMakeLists.txt:20]: We should leave the ZED and CUDA installation to the user. I will add a README to the ZED plugin on how to do this. We should add a section for ZED in the ILLIXR docs as well.

File: zed/CMakeLists.txt (hunk 1 above)
Comment (charmoniumQ) [reply zed/CMakeLists.txt:25]: Try building it without. We will remove if it succeeds.

File: zed/CMakeLists.txt (hunk 2 above)
Comment (charmoniumQ) [reply zed/CMakeLists.txt:41]: I removed this.

File: zed/Makefile [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+plugin.opt.so:
+	mkdir -p build/ && \
+	cd build/ && \
+	cmake .. -DDNDEBUG=1 -DCMAKE_BUILD_TYPE=Release .. && \
+	make "-j$(nproc)" && \
+	cd ../ && \
+	rm -f plugin.opt.so && \
+	ln -s build/libplugin.so plugin.opt.so && \
+	true
+
```
[the remark this answers is not in the document]
Comment (charmoniumQ) [reply zed/Makefile:22]: Added.

File: zed/src/main.cpp (hunk 6 above)
Comment (charmoniumQ) [reply zed/src/main.cpp:238]: Removed.

File: zed/src/main.cpp (hunk 5 above)
Comment (charmoniumQ) [reply zed/src/main.cpp:118]: Reverted.

File: zed/src/main.cpp (hunk 4 above)
Comment (charmoniumQ) [reply zed/src/main.cpp:81]: Removed.

File: zed/CMakeLists.txt (hunk 3 above)
Comment (mhuzai) [reply zed/CMakeLists.txt:20]: Why leave them up to the user? We install every other dependency for them, so why leave these two?

File: zed/CMakeLists.txt [hunk 8]
```diff
@@ -37,6 +37,8 @@ link_directories(${CUDA_LIBRARY_DIRS})
 FILE(GLOB_RECURSE SRC_FILES src/main.cpp)
 FILE(GLOB_RECURSE HDR_FILES include/*.hpp)
 
+ADD_EXECUTABLE(${libName} ${HDR_FILES} ${SRC_FILES})
```
Comment (mhuzai) [inline zed/CMakeLists.txt:39]: Why is this necessary?

File: zed/Makefile [hunk 9]
```diff
@@ -20,13 +20,13 @@ plugin.opt.so: build/Release/Makefile
 build/Debug/Makefile:
 	mkdir -p build/Debug && \
 	cd build/Debug && \
-	cmake -DCMAKE_BUILD_TYPE=Debug  -DCMAKE_CXX_COMPILER=$(CXX) -DCMAKE_C_COMPILER=$(CC) ../.. && \
+	cmake -DCMAKE_BUILD_TYPE=Debug  -DCMAKE_CXX_COMPILER=$(CXX) -DCMAKE_C_COMPILER=$(CC) -DCMAKE_LIBRARY_PATH=/usr/local/cuda/lib64/stubs ../.. && \
```
Comment (mhuzai) [inline zed/Makefile:23]: The user's CUDA may not be here. Furthermore, why is this required in the first place?

File: zed/README.md [hunk 10]
```diff
@@ -7,7 +7,8 @@
 
 ### Prerequisites
 
-- Windows 10, Ubuntu LTS
-- [ZED SDK](https://www.stereolabs.com/docs/installation/linux/) and its dependencies ([CUDA](https://developer.nvidia.com/cuda-downloads), [OpenCV](https://github.com/opencv/opencv/releases))
+- Ubuntu 18.04
+- [ZED SDK](https://www.stereolabs.com/docs/installation/linux/) 
+- [CUDA 10.2](https://developer.nvidia.com/cuda-toolkit-archive)
```
Comment (mhuzai) [inline zed/README.md:12]: CUDA 11 runs fine as well -- I have tested it. We shouldn't limit people to 10.2.

**Review by mhuzai — APPROVED** [review]

**Review by Hyjale — APPROVED** [review]
